src/snowflake/cli/api/metrics.py

```diff
@@ -0,0 +1,57 @@
+# Copyright (c) 2024 Snowflake Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Dict, Optional
+
+_FEATURES_PREFIX = "features"
+_APP_PREFIX = "app"
+
+
+class CLICounterField:
+    TEMPLATES_PROCESSOR = f"{_FEATURES_PREFIX}.templates_processor"
+    SQL_TEMPLATES = f"{_FEATURES_PREFIX}.sql_templates"
+    PDF_TEMPLATES = f"{_FEATURES_PREFIX}.pdf_templates"
+    SNOWPARK_PROCESSOR = f"{_FEATURES_PREFIX}.{_APP_PREFIX}.snowpark_processor"
+    POST_DEPLOY_SCRIPTS = f"{_FEATURES_PREFIX}.{_APP_PREFIX}.post_deploy_scripts"
+
+
+class CLIMetrics:
+    """
+    Class to track various metrics across the execution of a command
+    """
+
+    def __init__(self):
+        self._counters: Dict[str, int] = {}
+
+    def __eq__(self, other):
+        if isinstance(other, CLIMetrics):
+            return self._counters == other._counters
+        return False
```

Comment:
There's a test [here](https://github.com/snowflakedb/snowflake-cli/blob/05a668991202879107205991d80351e760f4eaf6/tests/api/test_global_cli_context.py#L48) that checks equality of the cli global context fields to make sure they were reset properly, and the default equality check wasn't passing since it just compares references